vm/src/deserializer.cpp

```diff
@@ -0,0 +1,42 @@
+#include "deserializer.h"
+#include <cassert>
+#include <stdexcept>
+
+namespace Borges {
+
+Deserializer::Deserializer(std::size_t n, const double* p)
+    : size_(n), data_(p), index_(0)
+{
+    assert(data_);
+}
+
+void
+Deserializer::deserialize(double* p)
+{
+    if (index_ == size_) {
+        throw std::runtime_error("buffer underflow");
+    }
+    double x = data_[index_];
+    ++index_;
+    *p = x;
+}
+
+void
+deserialize(Deserializer& d, bool* p)
+{
+    *p = deserialize<double>(d);
+}
```

Comment:
shouldn't this be deserialize<bool>(d) ?

Reply:
No, the template argument determines what type of value we should attempt to read from the buffer. Booleans are serialized as doubles, so they should be deserialized as doubles. The assignment to *p implicitly casts the double to bool.

Reply:
ah. in that case, an explicit cast is more clear.

Reply:
I disagree. It would also be redundant, since you'd still have to write:

    static_cast<double>(deserialize<double>(d))